pkg/graveler/ref/repository_iterator.go

```diff
@@ -4,99 +4,87 @@ import (
 	"github.com/treeverse/lakefs/pkg/graveler"
+	"github.com/treeverse/lakefs/pkg/kv"
 )
 
-type RepositoryIterator struct {
-	db        db.Database
-	ctx       context.Context
-	value     *graveler.RepositoryRecord
-	buf       []*graveler.RepositoryRecord
-	offset    string
-	fetchSize int
-	err       error
-	state     iteratorState
+type RepositoryIterator interface {
+	Next() bool
+	SeekGE(id graveler.RepositoryID)
+	Value() *graveler.RepositoryRecord
+	Err() error
+	Close()
 }
 
-func NewRepositoryIterator(ctx context.Context, db db.Database, fetchSize int) *RepositoryIterator {
-	return &RepositoryIterator{
-		db:        db,
-		ctx:       ctx,
-		fetchSize: fetchSize,
-		buf:       make([]*graveler.RepositoryRecord, 0, fetchSize),
-	}
+type KVRepositoryIterator struct {
+	ctx   context.Context
+	it    kv.MessageIterator
+	err   error
+	value *graveler.RepositoryRecord
+	store kv.Store
 }
 
-func (ri *RepositoryIterator) Next() bool {
-	if ri.err != nil {
-		return false
+func NewKVRepositoryIterator(ctx context.Context, store *kv.StoreMessage) (*KVRepositoryIterator, error) {
+	it, err := kv.NewPrimaryIterator(ctx, store.Store, (&graveler.RepositoryData{}).ProtoReflect().Type(), graveler.RepoPartition(), []byte(graveler.RepoPath("")), kv.IteratorOptionsAfter([]byte{}))
+	if err != nil {
+		return nil, err
 	}
+	return &KVRepositoryIterator{
+		it:    it,
+		store: store.Store,
+		ctx:   ctx,
+	}, nil
+}
 
-	ri.maybeFetch()
-
-	// stage a value and increment offset
-	if len(ri.buf) == 0 {
+func (ri *KVRepositoryIterator) Next() bool {
+	if ri.Err() != nil {
 		return false
 	}
-	ri.value = ri.buf[0]
-	ri.buf = ri.buf[1:]
-	ri.offset = string(ri.value.RepositoryID)
-	return true
-}
 
-func (ri *RepositoryIterator) maybeFetch() {
-	if ri.state == iteratorStateDone {
-		return
+	if !ri.it.Next() {
+		ri.value = nil
+		return false
 	}
-	if len(ri.buf) > 0 {
-		return
+	e := ri.it.Entry()
+	if e == nil {
+		ri.err = graveler.ErrNotFound
+		return false
 	}
 
-	var offsetCondition string
-	if ri.state == iteratorStateInit {
-		offsetCondition = iteratorOffsetCondition(true)
-		ri.state = iteratorStateQuerying
-	} else {
-		offsetCondition = iteratorOffsetCondition(false)
-	}
-	ri.err = ri.db.Select(ri.ctx, &ri.buf, `
-			SELECT id, storage_namespace, creation_date, default_branch
-			FROM graveler_repositories
-			WHERE id `+offsetCondition+` $1
-			ORDER BY id ASC
-			LIMIT $2`, ri.offset, ri.fetchSize)
-	if ri.err != nil {
-		return
-	}
-	if len(ri.buf) < ri.fetchSize {
-		ri.state = iteratorStateDone
+	repo, ok := e.Value.(*graveler.RepositoryData)
+	if repo == nil || !ok {
+		ri.err = graveler.ErrNotFound
+		return false
 	}
+
+	ri.value = graveler.RepoFromProto(repo)
+	return true
 }
 
-func (ri *RepositoryIterator) SeekGE(id graveler.RepositoryID) {
-	if errors.Is(ri.err, ErrIteratorClosed) {
+func (ri *KVRepositoryIterator) SeekGE(id graveler.RepositoryID) {
+	if errors.Is(ri.Err(), kv.ErrClosedEntries) {
```

Comment:
Question: Should we ignore seek on any iterator err, or continue with the same behavior as in the DB implementation?

Reply:
That is a good question.
I would vouche for keeping the same behavior, at least for the phase when we supporting both.
If you believe a change of behavior is beneficial, we can open an issue to handle it once transition is complete, and SQL support is abandoned.
WDYT?

Reply:
That's a good idea, we have several areas in the code that behave the same way. We should address all of them in the issue